### PR TITLE
feat: add computed workspace and agent health fields to the api

### PR DESCRIPTION
**Pull request by mafredri**
This PR adds support for workspace and agent health fields.

See: https://github.com/coder/coder/pull/8280#issuecomment-1620064738

<details>
<summary>Previous iterations</summary>
Example output for a workspace:

```json
{
    // ...
    "health": {
        "healthy": false,
        "failing_sections": [
            "agents.dbd7447f-e4e2-4d9f-92a2-136f84df4b2f"
        ],
        "agents": {
            "dbd7447f-e4e2-4d9f-92a2-136f84df4b2f": {
                "healthy": false,
                "reason": "agent startup script exited with an error"
            }
        }
    }
}
```

Update: We now only have a health field for agents to keep it simple:

```json
{
	// ...
	"health": {
        "healthy": false,
        "reason": "agent startup script exited with an error"
    }
}
```

To check on workspace health, one can do `const unhealthyAgents = workspace.latest_build.resources.map(r => r.agents).flat().filter(a => !a.health.healhty)`
</details>

Refs #6461

### CODE REVIEW

File: codersdk/workspaces.go
Comment (mtojek) [inline codersdk/workspaces.go:53]: API review: As Go maps don't keep the order, do you think that we should replace it with a corresponding slice?

File: codersdk/workspaces.go
Comment (mtojek) [inline codersdk/workspaces.go:51]: This will work until we've only `healthy`/`unhealthy`. What if we decide to add more states, for instance:`unknown`, `recovering`, `zombie`, `tainted`, etc.

Maybe we should replace the `healthy` with `state: healthy`?

File: coderd/workspaceagents.go
Comment (mtojek) [inline coderd/workspaceagents.go:1267]: I might be not in the loop, but what is the reason for returning the human-readable reason over API? Did you consider putting this mapping in the site/UI code?

Let's say that we want the site/UI to take an action on `agent is taking too long to connect`, does it mean that we have to compare strings?

**Review by mtojek — COMMENTED** [review]
I may not see the full picture, but if we want to KISS, why just not add 2 properties for every agent:

```
"health": {
    "healthy": false,
    "reason": "Foobar..."
}
```

I have a feeling that `failing_sections` seems to be an overkill for what we're trying to achieve.

File: codersdk/workspaces.go
Comment (mafredri) [reply codersdk/workspaces.go:53]: True, maintaining order might be worthwhile, but we'll have to include the agent ID in `WorkspaceAgentHealth` which is duplicated information on the agent endpoints. Perhaps not a big deal though. Failing sections could still work with a slice given that we would use index (e.g. `['agents.0']`).

File: codersdk/workspaces.go
Comment (mafredri) [reply codersdk/workspaces.go:51]: Based on https://github.com/coder/coder/issues/6461#issuecomment-1612826927, we don't want more values in the enum. All of the mentioned states (`unknown`, `recovering`, ...) would work under `unhealthy` with the reason field filled out (or similar).

File: codersdk/workspaces.go
Comment (mtojek) [reply codersdk/workspaces.go:53]: Usually it's annoying when you're refreshing the API and order of items is shuffled every time, so that's the main reason of my recommendation.

File: coderd/workspaceagents.go
Comment (mafredri) [reply coderd/workspaceagents.go:1267]: This is a valid concern. I suppose this method allows us to skip the introduction of a new enum and simply surface the reason in the UI. This is similar to how we do with API errors currently where we have a human readable reason in the HTTP response.

If we make this an enum, it needs to be kept up-to-date over UI, CLI, etc.

**Comment by mtojek** [comment]
> it's nice to be able to check one property on the workspace instead of diving into all the agents to figure out where the issue is. 

Let me challenge this.

Coder workspaces use at most 1 (maybe 2?) agent per workspace. It does not require complex UI processing on the frontend side. Also, I would consider deferring `failing_sections` for later until we have reasons to add the property. I would not build a super-flexible endpoint if we don't have plans to use the field. More fields = more maintenance. 

> This is simplified doesn't work as-is on the agents map (agents.uuid key) without extra processing, but hopefully illustrates the use-case.

I would not drive the API design with front-end shortcuts. What is important for a good API design is making endpoints concise and idempotent. IMHO`failing_sections` is definitely redundant, and I would rather treat it as a field for future optimizations.

**Comment by mafredri** [comment]
> I would not drive the API design with front-end shortcuts. What is important for a good API design is making endpoints concise and idempotent. IMHO`failing_sections` is definitely redundant, and I would rather treat it as a field for future optimizations.

I do see your point, however, I would rather equate this with restful APIs where e.g. the URL to the "next page of results" is part of the response. Likewise this `failing_sections` field is pointing towards what caused the workspace to become unhealthy.

That said, I'm fine with removing the sections feature. The tradeoff is that if we do add more reasons for unhealthiness, we'll need to update all the places that only took agents into consideration (for better or worse).

**Comment by mtojek** [comment]
> That said, I'm fine with removing the sections feature. The tradeoff is that if we do add more reasons for unhealthiness, we'll need to find and update all the places that only took agents into consideration (for better or worse).

Thanks for addressing the issue. I think that this form is still open for modifications unless we predict other health factors to consider?

Anyway, as chatted with you over slack, I don't intend to slow down this feature delivery, so if more voices are for keeping the `failing_sections` metadata, I will not block this PR. I just don't see this to be super-complicated now, so that we need an extra helper field to "calculate" workspace health.

**Comment by mafredri** [comment]
@mtojek thinking about this some more, how do you feel about the following format for the API response?

```json
{
    // ...
    "health": {
        "healthy": false,
        "failing_agents": ["dbd7447f-e4e2-4d9f-92a2-136f84df4b2f"]
    },
    "resources": [
        {
            // ...
            "agents": [
                {
                    "id": "dbd7447f-e4e2-4d9f-92a2-136f84df4b2f",
                    // ...
                    "health": {
                        "healthy": false,
                        "reason": "timeout|disconnected|start_error|..."
                    }
                }
            ]
        }
    ]
}
```

We've eliminated the duplication but kept the response informative. This can't be extended in a generic way in the future, so if we want more knobs for health (like workspace build failures, etc.), we need to refactor the code that uses it. That should be fairly simple, though.

From a client/UI perspective this is a bit harder to use than duplicating the information in the workspace health, but if we're fine with that, we can use this approach.

Whether or not `reason` should be human readable or an enum (enum of the "bad" states for `status|lifecycle_state`), I'll leave up to @BrunoQuaresma and @bpmct.

PS. Maybe we should make `health` a property of workspace build (`latest_build`) instead of the workspace itself? Not sure... conceptually both work.

**Comment by mtojek** [comment]
I like this approach more. It is intuitive, and a machine reader won't be required to process the details of the agent structure. In theory, we don't expect our APIs to be human-readable, but I don't know if this requirement changes in the future.

> PS. Maybe we should make health a property of workspace build (latest_build) instead of the workspace itself? Not sure... conceptually both work.

Workspace build is a process of provisioning the workspace, so it can't expose any health metrics by its nature 🤔. I would rather stick to the workspace health.

File: coderd/workspaces_test.go
Comment (mtojek) [inline coderd/workspaces_test.go:206]: nit: assert empty for `failingAgents`, right?

File: coderd/workspaces_test.go
Comment (mtojek) [inline coderd/workspaces_test.go:249]: Maybe one more test case with multiple agents to check the logic?

**Review by mtojek — APPROVED** [review]
Awesome 👍

File: coderd/workspaces_test.go
Comment (mafredri) [reply coderd/workspaces_test.go:206]: Good catch! 👍🏻

**Review by BrunoQuaresma — APPROVED** [review]
LGTM, we can use ENUM or human-readable messages. I don't have thoughts about this... I already worked with both and never saw a HUGE difference in development or maintainability in the long term so whatever is best for the BE.

**Comment by mafredri** [comment]
> LGTM, we can use ENUM or human-readable messages. I don't have thoughts about this... I already worked with both and never saw a HUGE difference in development or maintainability in the long term so whatever is best for the BE.

Alright, let's go with this format for now 👍🏻. If we want an enum instead, we can introduce it later or add a second field (`reason_code`, etc.).

**Pull request merged**
